cupy/core/core.pyx

```diff
@@ -167,6 +167,21 @@ cdef class ndarray:
             desc['data'] = (self.data.ptr, False)
         else:
             desc['data'] = (0, False)
+        if _util.CUDA_ARRAY_INTERFACE_SYNC:
+            stream_ptr = stream_module.get_current_stream_ptr()
+            # TODO(leofang): check if we're using PTDS
+            # CAI v3 says setting the stream field to 0 is disallowed
+            if stream_ptr == 0:
+                desc['stream'] = 1  # TODO(leofang): use runtime.streamLegacy
+            else:
+                desc['stream'] = stream_ptr
+        else:
+            # Old behavior (prior to CAI v3): stream sync is explicitly handled
+            # by users. To restore it, we do not export any stream if
+            # CUPY_CUDA_ARRAY_INTERFACE_SYNC is set to 0 (so that other
+            # participating libraries lacking a finer control over sync
+            # behavior can avoid syncing).
+            desc['stream'] = None
```

Comment:
If this is set to `None` then it's implied that `self.data.ptr` can be used on any stream safely. I'm not 100% familiar with cupy's execution patterns, but would we need to synchronize here to guarantee that?

Reply:
If a user sets `CUPY_CUDA_ARRAY_INTERFACE_SYNC`, IMHO he/she should get the full control over the synchronization. First, a nice thing CuPy has, unlike the notorious TensorFlow for example, is there's no internal streams floating around, so all CuPy streams are explicitly managed by users, and without specifying any stream the default in CuPy is `cupy.cuda.Stream.null` (up to this PR; see the ongoing discussion in #4322).

Moreover, an ability I'd like to have is a way to "forget" the previous stream. Suppose I "launder" the array over 3 libraries (say, CuPy -> mpi4py -> Numba), since mpi4py does not explicitly handle any CUDA stuff (it's delegated to the MPI library), by the time a CuPy array arrives at Numba it still holds the stream from CuPy if we didn't empty it, and Numba could synchronize over it, which is semantically incorrect. What should have been done at the transition points (by the user) is simply
1. before entering mpi4py: to sync the stream (if any),
2. before entering Numba: do either nothing (for a blocking MPI call) or call `MPI.Wait()` (for a non-blocking MPI call),

all of which need to be performed explicitly by the user. Then, the stream in CAI is irrelevant in this picture.

Reply:
This is then violating the specification: https://github.com/numba/numba/blob/master/docs/source/cuda/cuda_array_interface.rst#interface-requirements

> - No synchronization is required on the part of the Consumer.
> - The Consumer may enqueue operations on the underlying data immediately on any stream.

Reply:
Hi @kkraus14 Recall that we agreed upon a few "backdoor" clauses 😁 For example, what's implemented here is to provide options for flexibility:
> Optionally, to facilitate the User relaxing conformance to synchronization semantics:
> - Producers may provide a configuration option to always set stream to None.
> - Consumers may provide a configuration option to ignore the value of stream and act as if it were None or not provided.

See https://numba.readthedocs.io/en/latest/cuda/cuda_array_interface.html#interface-requirements (need to scroll down a bit).

For now, this PR uses `CUPY_CUDA_ARRAY_INTERFACE_SYNC` to control both Producer and Consumer behaviors. Perhaps I could introduce another env var, say, `CUPY_CUDA_ARRAY_INTERFACE_NO_STREAM_EXPORT` for controlling Producer (and let the other env var controlling Consumer)? Would this be better?

Finally, let me reiterate that without Users explicitly setting the env var(s), my PR is strictly compliant with the v3 protocol by default. Without making sure compliance our time invested in the Numba CAI v3 PR would a waste, after all 🙂

Reply:
Sounds good, thanks for the explanation!

Reply:
Done in 56ede6d. I used a positive term `CUPY_CUDA_ARRAY_INTERFACE_EXPORT_STREAM` as I want both env vars to default to 1.

Reply:
I feel `CUPY_CUDA_ARRAY_INTERFACE_VERSION` (defaults to `3`) is better. How do you think?

Reply:
Hi @asi1024 For this and your other comment, I have a mixed feeling. On the one hand I totally get that you want a flag to say we're complaint up to v3. On the other hand, the expectation for `!=3` is unclear to me (for example, do we error out when a foreign library is on v3 but we're set to v2, or vice versa?), and it's a little less intuitive for Users to know the meaning of `CUPY_CUDA_ARRAY_INTERFACE_VERSION` vs `CUPY_CUDA_ARRAY_INTERFACE_EXPORT_STREAM`/`CUPY_CUDA_ARRAY_INTERFACE_SYNC`. Finally, do we use the same env var to control the behavior of both Producer and Consumer?

Note all the fields added since v0 are optional, so from v0 to v3 didn't really change much, except that the safety of operation is thoroughly considered. Also, AFAIK no other participating libraries check the `version` field.

From my standpoint as long as both of them can be set to opt out by Users, I don't have strong opinion. But I do want to make sure that everyone here is happy with this PR, and that we don't make mistakes that we have to revisit later, as this PR will be backported to a stable release (v8.3?) and we better not mess up 🙂

Reply:
May `CUPY_CUDA_ARRAY_INTERFACE_EXPORT_VERSION` be much better?

Reply:
Hi @asi1024 Do you mean we just rename `CUPY_CUDA_ARRAY_INTERFACE_EXPORT_STREAM` to `CUPY_CUDA_ARRAY_INTERFACE_EXPORT_VERSION`, or you have something more in mind?

Reply:
Yes! I mean just renaming the variable.

Reply:
Ahhh OK I totally misunderstood 😅 Thanks for bearing with me...

Reply:
@asi1024 Sorry I am confused again...What would be the allowed values for `CUPY_CUDA_ARRAY_INTERFACE_EXPORT_VERSION`? Only 2 and 3 (default)? Do we also need to fix the value for the key `version`?

Reply:
Yes, only 2 and 3 are supported, and an error will be raised otherwise. The key `version` also need to be fixed.

Reply:
Thanks, @asi1024, I updated it. PTAL.